### PR TITLE
Generalize to allow Integer flow

### LINKED ISSUE #3401: Integer flow variables?
Currently flow-variables cannot be integers, which seems a bit odd:

https://github.com/modelica/ModelicaSpecification/blob/07ccecd0719e130ad7edfb578f94e72ba5872dbe/chapters/classes.tex#L236

especially since integers do form an additive group. The restriction was added in Modelica 1.1 and I believe we should lift that restriction.
(Note that Boolean and Strings do _not_ form an additive group so we shouldn't generalize it further.)

Use-cases:

- Counting the number of instances of a class (using inner/outer).
- The flow is just an integer; so better to handle it in that way.

Implementation effort:

- Minimal: It seems it is just a matter of removing a restriction - as the resulting equations can be treated as any other equations.
- Dymola currently generates a diagnostics and then handles it.

### PR DESCRIPTION
**Pull request by HansOlsson**
Closes #3401 
Note that the preceding text in the DAE-chapter already mention that it is sorted to ensure that the variable to solve for is unique.

### CODE REVIEW

**Comment by HansOlsson** [comment]
Would be good with some reviews on this one. The change (apart from the flow semantics) is that it will allow slightly more general integer equations, but still restricted to "almost trivial" integer equations.

File: chapters/dae.tex [hunk 1]
```diff
@@ -81,7 +81,13 @@ \chapter{Modelica DAE Representation}\label{modelica-dae-representation}
 
 The key difference between the two groups of discrete-time variables $z$ and and $m$ here is how they are determined.
 The interpretation of the solved form of \eqref{eq:dae-discrete-valued} is that given values for everything else, there is a closed-form solution for $m$ in the form of a sequence of assignments to each of the variables of $m$ in turn -- there must be no cyclic dependencies between the equations used to solve for $m$.
-Further, each of the original model equations behind \eqref{eq:dae-discrete-valued} must be given in solved form, at most requiring flipping sides of the equation to obtain the used assignment form.
+Further, each of the original model equations behind \eqref{eq:dae-discrete-valued} must be given in almost solved form:
+\begin{itemize}
+\item
+  For non-\lstinline!Integer! at most requiring flipping sides of the equation to obtain the used assignment form.
```
Comment (henrikt-ma) [inline chapters/dae.tex:87]: More readable:
```suggestion
  Non-\lstinline!Integer! equations at most requiring flipping sides of the equation to obtain the used assignment form.
```

File: chapters/dae.tex [hunk 2]
```diff
@@ -81,7 +81,13 @@ \chapter{Modelica DAE Representation}\label{modelica-dae-representation}
...
+\item
+  For \lstinline!Integer! uniquely as a term (without any multiplicative factor) on either side of the equation, at most requiring addition or subtraction of a previously computed value to obtain the used assignment form.
```
Comment (henrikt-ma) [inline chapters/dae.tex:89]: Match style of previous item:
```suggestion
  For \lstinline!Integer! equations the solved variable must appear uniquely as a term (without any multiplicative factor) on either side of the equation, at most requiring addition or subtraction of other terms in the equation to obtain the used assignment form.
```

**Review by henrikt-ma — CHANGES_REQUESTED** [review]
Basically good.  Only suggesting some rewording.

**Comment by HansOlsson** [comment]
> Basically good. Only suggesting some rewording.

Done.

**Review by henrikt-ma — COMMENTED** [review]
Looks good to me.  However, having added seemingly simple changes to the specification too quickly in the past, I think we should make sure this has been properly test implemented before merging.

**Review by henrikt-ma — APPROVED** [review]

**Comment by henrikt-ma** [comment]
I have now made a test implementation, and realize that the main challenge here is to define the required form of the equations where an `Integer` variable can be solved, so that tools have a fair chance of implementing this consistently.

To illustrate, consider various extensions of this base model:
```
partial model IntegerSolvabilityBase
  Integer x;
  Integer y;
equation
  x + y = 0;
end IntegerSolvabilityBase;
```

It's not too difficult to support this valid case:
```
model ValidSolvability
  extends IntegerSolvabilityBase;
equation
  y + 1 = 2;
end ValidSolvability;
```

Also, it is not difficult to reject this:
```
model ClearlyNotSolvable
  extends IntegerSolvabilityBase;
equation
  1.5 * y + 1 = 4;
end ClearlyNotSolvable;
```

I was delighted to find that my test implementation also rejects this:
```
  1.0 * y + 1 = 4;
```

It also rejects this:
```
  2 * y + 1 = 5;
```
and this:
```
  2 * (y + 1) = 0;
```

However, I was not surprised to find that it doesn't reject this:
```
  1 * y + 1 = 4;
```
nor this:
```
  y + y + 1 - y = 0;
```

To make it even more complicated, we can mix in an evaluated parameter:
```
model EvaluatedParameter
  extends IntegerSolvabilityBase;
  parameter Integer p = 1;
  Real[p] z = fill(0.0, p);
equation
  p * y + 1 = 4;
end EvaluatedParameter;
```

The possibilities for variation of examples that could be difficult to reject according to the currently proposed restrictions are endless.  The line between correctly rejected and incorrectly passing models appears to a large degree to be an arbitrary outcome of what sort of simplifications that are preformed before the solvability of a variable in an equation is determined.  For me, it is very difficult to know which of these simplifications that are needed to make other models out there translate as expected, and I also believe that we want the specification to leave room for tools to do different simplifications/optimizations.  

I'm not saying it _has to be_ impossible to guard simplifications/optimizations in the front end, so that they are not applied if they would change the solvability of an `Integer` variable, but I see a clear risk that it could be a major effort to put all guards in place, and that I can't predict whether the guards can be added without also causing problems for existing libraries.

Is Dymola better at rejecting the invalid ones than my test implementation?  Do we really want to have a definition which seems so hard to implement correctly, or should we look for alternative formulations that would better match what can be implemented (more) consistently without too much effort?

**Comment by henrikt-ma** [comment]
@HansOlsson, is it intentional to leave the `Real` requirement for the `flow` variable of a stream connector?

**Comment by HansOlsson** [comment]
> @HansOlsson, is it intentional to leave the `Real` requirement for the `flow` variable of a stream connector?

Yes, the explanation takes the limit as the flow goes to zero - and a discrete-valued type like Integer doesn't work with that.

**Comment by HansOlsson** [comment]
> Is Dymola better at rejecting the invalid ones than my test implementation? 

No.

>Do we really want to have a definition which seems so hard to implement correctly, or should we look for alternative formulations that would better match what can be implemented (more) consistently without too much effort?

Ideally, but to me those test-cases are quite esoteric, and at least in Dymola the same logic that handles these could sometimes handle the previous formulation; so we are not making the specification worse by this change.

**Comment by henrikt-ma** [comment]
> Ideally, but to me those test-cases are quite esoteric, and at least in Dymola the same logic that handles these could sometimes handle the previous formulation; so we are not making the specification worse by this change.

Yes, and let me just clarify this point with an example.  Which tools are currently able to reject the following?
```
model InvalidIntegerEquation
  Integer k;
equation
  1 * k = 2;
end InvalidIntegerEquation;
```

**Pull request merged**